firmware/console/binary/tunerstudio_io.cpp

```diff
@@ -245,20 +242,34 @@ void sr5WriteCrcPacket(ts_channel_s *tsChannel, const uint8_t responseCode, cons
 	}
 #endif /* TS_CAN_DEVICE */
 
-	*(uint16_t *) writeBuffer = SWAP_UINT16(size + 1);   // packet size including command
-	*(uint8_t *) (writeBuffer + 2) = responseCode;
+	auto scratchBuffer = tsChannel->crcReadBuffer;
 
-	// CRC on whole packet
-	uint32_t crc = crc32((void *) (writeBuffer + 2), 1); // command part of CRC
-	crc = crc32inc((void *) buf, crc, (uint32_t) (size)); // combined with packet CRC
+	// don't transmit a null buffer...
+	if (!buf) {
+		size = 0;
+	}
 
-	*(uint32_t *) (crcBuffer) = SWAP_UINT32(crc);
+	// don't transmit too large a buffer
+	efiAssertVoid(OBD_PCM_Processor_Fault, size <= BLOCKING_FACTOR + 7, "sr5WriteCrcPacket tried to transmit too large a packet")
 
-	sr5WriteData(tsChannel, writeBuffer, 3);      // header
-	if (size > 0) {
-		sr5WriteData(tsChannel, (const uint8_t*)buf, size);      // body
+	// If transmitting data, copy it in to place in the scratch buffer
```

Comment:
shall we change to "If transmitting data, copy it in to place in the scratch buffer to make sure that we calculate CRC of the immutable copy"

Reply:
good idea!